### ISSUE TITLE
[FSDP] UnpicklingError when calling save_state_dict in distributed run

### ISSUE THREAD

**Issue by shijie-wu** [pytorch/pytorch#109675]
### 🐛 Describe the bug

we are using multi-nodes training with FSDP, and we got the following error during checkpointing through `torch/distributed/checkpoint/state_dict_saver.py`
```
File "/opt/ml/code/checkpoints.py", line 153, in _save
checkpoint.save_state_dict(
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/checkpoint/state_dict_saver.py", line 119, in save_state_dict
return distW.all_reduce("write", write_data, finish_checkpoint)
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/checkpoint/utils.py", line 236, in all_reduce
final_result = self.broadcast_object(result)
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/checkpoint/utils.py", line 88, in broadcast_object
dist.broadcast_object_list(
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/distributed_c10d.py", line 1451, in wrapper
return func(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/distributed_c10d.py", line 2277, in broadcast_object_list
object_list[i] = _tensor_to_object(obj_view, obj_size)
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/distributed_c10d.py", line 1970, in _tensor_to_object
return _unpickler(io.BytesIO(buf)).load()
_pickle.UnpicklingError: invalid load key, '\x00'.
```
originally, we thought that it only happened when we overwrite checkpoint to an existing folder, but we recently observed the error occurred even when writing to fresh folder. thus this error is non-deterministic. it might be related to these issues https://github.com/pytorch/pytorch/issues/109396 

### Versions

```
Collecting environment information...
PyTorch version: 2.0.1
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A
```

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin

**Comment by XilunWu** [comment]
Hi @shijie-wu I saw #109396 was solved. Does that help you with this issue? If not, could you try updating NCCL to 2.17.3 or above and see if the issue is gone?

**Comment by XilunWu** [comment]
@fegin again this seems to be the same problem that we had 2 weeks ago:
1. nccl `broadcast_object_list` has pickling issue while gloo doesn't
2. `size` is broadcasted correctly while the `object` is wrong

However, switching to NCCL 2.18.3 does not solve the issue. Let's discuss offline to see what we can do to help.

**Comment by rahul003** [comment]
Any update from your discussion @XilunWu ?

**Comment by viclzhu** [comment]
Yes, the root cause for us actually appeared to be not setting `export NCCL_PROTO="simple"`. EFA does not support `LL/LL128`, but without setting that, would use it, causing the silent data corruption. For us this was not automatically set with `aws-ofi-nccl==1.5.0`, but in later versions it should automatically be set during runtime.

The workaround would be to either manually set the env var, or upgrade to a newer `aws-ofi-nccl` version that sets it automatically.
